File: java-src/Sorts.java
/**
 * Java sorting algorithms
 *
 * @author Louie Brooks
 * Written on: starting 07/02/24
 */

public class Sorts {

    /**
     * Runs bubble sort on array of integers
     *
     * @param data
     * @return Ascending sorted array
     */
    public static int[] bubble(int[] data) {

        // inter-pass loops
        for (int i = 0; i < data.length - 1; i++) {

            boolean swapped = false;

            // intra-pass loops
            for (int j = 1; j <= data.length - 1 - i; j++) {

                // swap if needed
                if (data[j] < data[j - 1]) {

                    int temp = data[j];
                    data[j] = data[j - 1];
                    data[j - 1] = temp;
                    swapped = true;

                }

            }

            if (!swapped) {
                break;
            }

        }

        return data;
    }

    /**
     * Runs merge sort on integer array
     *
     * @param data
     * @return Ascending sorted array
     */
    public static int[] merge(int[] data, int start, int end) {

        // recursively partition until each list is 1 element long
        if (start == end) {
            int[] singleElement = new int[1];
            singleElement[0] = data[start];
            return singleElement;
        }

        // split list into 2
        int mid = (int) Math.floor((start + end) / 2.0);
        int[] leftArr = merge(data, 0, mid);
        int[] rightArr = merge(data, mid + 1, end);

        return merge2Lists(leftArr, rightArr);

    }

    public static int[] merge2Lists(int[] arr1, int[] arr2) {

        int pointer1 = 0;
        int pointer2 = 0;
        int pointerMerged = 0;

        int[] mergedArray = new int[arr1.length + arr2.length];

        while (pointer1 < arr1.length && pointer2 < arr2.length) {

            if (arr1[pointer1] > arr2[pointer2]) {
                mergedArray[pointerMerged] = arr2[pointer2];
                pointer2++;
                pointerMerged++;
            }

            else {
                mergedArray[pointerMerged] = arr1[pointer1];
                pointer1++;
                pointerMerged++;
            }
        }

        while (pointer1 < arr1.length) {
            mergedArray[pointerMerged] = arr1[pointer1];
            pointer1++;
            pointerMerged++;
        }

        while (pointer2 < arr2.length) {
            mergedArray[pointerMerged] = arr2[pointer2];
            pointer2++;
            pointerMerged++;
        }

        return mergedArray;

    }

    /**
     * Runs insertion sort on integer array
     *
     * @param data
     * @return Ascending sorted array
     */
    public static int[] insertion(int[] data) {

        for (int j = 1; j < data.length; j++) {
            int key = data[j];
            int i = j - 1;
            while (i >= 0 && data[i] > key) {
                data[i + 1] = data[i];
                i = i - 1;
            }
            data[i + 1] = key;
        }

        return data;
    }

    /**
     * Runs selection sort on integer array
     *
     * @param data
     * @return Ascending sorted array
     */
    public static int[] selection(int[] data) {

        for (int lastUnsorted = data.length - 1; lastUnsorted > 0; lastUnsorted--) {

            int positionOfLargest = lastUnsorted;

            for (int i = 0; i < lastUnsorted; i++) {

                if (data[i] > data[positionOfLargest]) {
                    positionOfLargest = i;
                }

            }

            if (positionOfLargest != lastUnsorted) {
                int temp = data[positionOfLargest];
                data[positionOfLargest] = data[lastUnsorted];
                data[lastUnsorted] = temp;
            }

        }

        return data;
    }

    /**
     * Outputs the elements of an array as a string
     *
     * @param arr
     * @return Elements of array as a string
     */
    public static String printArrContents(int[] arr) {

        String message = "";

        for (int i = 0; i < arr.length; i++) {
            message += arr[i];
        }

        return message;
    }

    /**
     * Test case for differnet sorting algorithms
     *
     * @param args
     */
    public static void main(String[] args) {

        int[] testData = {9, 8, 7, 6, 5, 4, 3, 2, 1};

        System.out.print("Bubble Sort: ");
        System.out.println(Sorts.printArrContents(Sorts.bubble(testData)));
        System.out.print("Merge Sort: ");
        System.out.println(Sorts.printArrContents(Sorts.merge(testData, 0, testData.length - 1)));
        System.out.print("Insertion Sort: ");
        System.out.println(Sorts.printArrContents(Sorts.insertion(testData)));
        System.out.print("Selection Sort: ");
        System.out.println(Sorts.printArrContents(Sorts.selection(testData)));

    }
}

File: pythonSrc/Sorts.py
# Python sorting algorithms
# Author: Louie Brooks
# Written on: starting 13/02/24

# imports
import math

# runs bubble sort and returns the sorted list
def bubble(data : list) -> list:
    
    for i in range(0, len(data) - 1):
        
        swapped = False

        for j in range(1, len(data) - i):

            if data[j] < data[j-1]:
                temp = data[j]
                data[j] = data[j-1]
                data[j-1] = temp
                swapped = True
        
        if not(bool(swapped)):
            break

    return data

# runs merge sort and returns the sorted list
def merge(data : list, start : int, end : int) -> list:

    if start == end:
    
        singleElement = [0]
        singleElement[0] = data[start]
    
        return singleElement
    
    mid = math.floor((start + end)/2)

    leftArr = merge(data, start, mid)
    rightArr = merge(data, mid + 1, end)

    return merge2Lists(leftArr, rightArr)

# merge the two lists provided in to one ascendling ordered array 
def merge2Lists(arr1 : list, arr2 : list) -> list:
    pointer1 = 0
    pointer2 = 0
    pointerMerged = 0

    mergedArray = [0]*(len(arr1) + len(arr2))

    while pointer1 < len(arr1) and pointer2 < len(arr2):

        if arr1[pointer1] > arr2[pointer2]:
            mergedArray[pointerMerged] = arr2[pointer2]
            pointer2 += 1
            pointerMerged += 1

        else:
            mergedArray[pointerMerged] = arr1[pointer1]
            pointer1 += 1
            pointerMerged += 1

    while pointer1 < len(arr1):
        mergedArray[pointerMerged] = arr1[pointer1]
        pointer1 += 1
        pointerMerged += 1

    while pointer2 < len(arr2):
        mergedArray[pointerMerged] = arr2[pointer2]
        pointer2 += 1
        pointerMerged += 1

    return mergedArray 

# runs insertion sort and returns the sorted list
def insertion(data : list) -> list:
    for i in range (1, len(data)):
        key = data[i]
        j = i - 1
        while j >= 0 and data[j] > key:
            data[j + 1] = data[j]
            j = j - 1
        data[j+1] = key
    return data

# runs selection sort and returns the sorted list
def selection(data):
    pass

# returns the elemetns of an array as a concanated string
def arrayToString(arr):
    message = ""

    for i in arr:
        message += str(i)

    return message
        
# runs the main code
testData = [9,8,7,6,5,4,3,2,1]
print("Bubble Sort: " + arrayToString(bubble(testData)))

# have to redefine the array as the methods currently sort it and treat lists as objects in python
testData = [9,8,7,6,5,4,3,2,1]
print("Merge Sort: " + arrayToString(merge(testData, 0, len(testData) - 1)))

# have to redefine the array as the methods currently sort it and treat lists as objects in python
testData = [9,8,7,6,5,4,3,2,1]
print(f"Insertion Sort: {arrayToString(merge(testData, 0, len(testData) - 1))}")